### PR TITLE
Proposal to strengthen passwords

### PR DESCRIPTION
**Pull request by ryanfeeley**
Make suggestions, and block known bad passwords

### CODE REVIEW

File: features/FxA-97-password-strength/README.md [hunk 1]
```diff
@@ -0,0 +1,67 @@
+# Strengthening Passwords
+
+## Problem Summary
+
+[comment]: # (Talk about current goals of the system and how they aren't being met. What outcomes are we not delivering? This can also include an explicit request for improvement that doesn't dictate a specific or concrete solution.)
```
Comment (shane-tomlinson) [inline features/FxA-97-password-strength/README.md]: Woah, I didn't know comments were possible like this.

File: features/FxA-97-password-strength/README.md [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+
+Users possess varying levels of understanding of what they are signing up for, and what is at risk. When syncing Firefox data (namely passwords) to a cloud-based account, reusing a password for their account puts them at greater risk to online attacks.
+
+### Assumptions
+
+[comment]: # (This is where you talk about what you assume to be true. This could include assumptions around what users are doing, errors, gaps, etc., based on anecdotes, opinions, testing, or data.)
+
+Since account registration has appeared during First Run in an iframe, we are receiving are more users, but with less of an understanding of the significance of making an account. Since the Choose What To Sync screen has launched, we've noticed that First Run users are twice as likely to abandon during sign-up than users from other entry points.
```
Comment (shane-tomlinson) [inline features/FxA-97-password-strength/README.md]: `receiving are` => `receiving`

File: features/FxA-97-password-strength/README.md [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+
+## Outcomes
+
+[comment]: # (What are the outcomes you want to achieve? What is the success criteria?)
+
+We feel that we could better serve user security and our [Data Privacy Principles](https://www.mozilla.org/en-US/privacy/principles/) by being more transparent about the risks and benefits of signing up for an account. This will hopefully result in stronger passwords as measured by our bloom filter, and also result in fewer account sign-ups (as many users will realize it's not something they need or want). Another goal is increasing multi-device usage, which in turn should result in an increase in overall Firefox engagement.
```
Comment (shane-tomlinson) [inline features/FxA-97-password-strength/README.md]: The second goal (increasing multi-device usage) seems orthogonal to measuring password strength unless messaging about using Fx on other devices is also included.

File: features/FxA-97-password-strength/README.md [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+
+## Hypothesis
+
+[comment]: # (A high level hypothesis of how the feature you're proposing is going help us achieve the outcomes listed above. I recommend this be a sentence of the form. We believe that doing this building this creating this experience for these people will achieve this outcome. We will know this is true when we see this qualitative feedback quantitative feedback KPI change.)
+
+We believe that providing more information about password strength for registering users will achieve stronger passwords. We will know this is true when we see a reduction in our bloom filter numbers.
```
Comment (shane-tomlinson) [inline features/FxA-97-password-strength/README.md]: `a reduction in our bloom filter numbers` is vague, and I think possibly incorrect. 

I theorize many users will enter an initial weak password, hit "submit", our password checker will flag the password, we will show the user some messaging, and many users will attempt a stronger password. If the user enters a weak password again, then our password strength checker "known weak password" count will increase, not decrease.

If we allow weak passwords to be used after notifying the user, I think the concrete measurement would be: a reduction in the % of users that create accounts with known weak passwords.

If we block weak passwords all together, I think a concrete measurement could be: % of users we prevented from signing up with known weak passwords.

File: features/FxA-97-password-strength/README.md [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+We believe that blocking known weak passwords will achieve stronger passwords. We will know this is true when we see a reduction in our bloom filter numbers.
+
+We believe that these changes will result in us collecting only the data that benefits users. We will know this is true when we see a slight reduction in the number of new accounts.
+
+## Metrics
+
+[comment]: # (How are you going to measure the outcome / success? Please provide sample artifact graphs here.)
+
+* Datadog Dashboard › FxA Content Server - Password Strength › % of users that hit each restriction
+ * Fewer users hitting ALL_LETTERS_OR_NUMBERS and BLOOMFILTER_HIT
```
Comment (shane-tomlinson) [inline features/FxA-97-password-strength/README.md]: Again, I think we may see an *increase* in the number of users hitting these flags. I think the important outcome is fewer users who create accounts with a password that hits one of these two flags.

File: features/FxA-97-password-strength/README.md [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+
+## Detailed design
+
+[comment]: # (This is the bulk of the RFC. Explain the design in enough detail for somebody familiar with the language to understand. This should get into specifics and corner-cases, and include examples of how the feature is used.)
+
+![Password Strength](password-strength.png)
+
+#### Password field give focus
+When the user gives focus to the password field on the registration form, show instructions below that say:
+* A strong, unique password will keep your Firefox data safe from intruders. <u>More info</u>.
```
Comment (shane-tomlinson) [inline features/FxA-97-password-strength/README.md]: Of course, this text here could [negate the increase I think could happen](https://github.com/mozilla/fxa/pull/150/files#r64023087).

File: features/FxA-97-password-strength/README.md [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+
+"More info" will link to a modal dialog containing concise instructions on creating a strong password based on [this SUMO article](https://support.mozilla.org/en-US/kb/create-secure-passwords-keep-your-identity-safe)
+
+#### User enters characters
+When the user enters any characters, the instructions change to become:
+* 8 characters minimum, but longer if you plan to sync passwords. <u>More info</u>.
```
Comment (shane-tomlinson) [inline features/FxA-97-password-strength/README.md]: `but longer if you plan to sync passwords` sounds like we are going to enforce the `longer if you plan to sync passwords` part. Perhaps something like `8 characters minimum, but we recommend longer if you plan to sync passwords.`

File: features/FxA-97-password-strength/README.md [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+
+#### Submitted password is on our blocklist
+When the user enters a password that is on our blocklist, the UI will behave as if the password field were empty, and show the following error:
+* This is a common password; please consider another one.  <u>More info</u>.
```
Comment (shane-tomlinson) [inline features/FxA-97-password-strength/README.md]: Are we going to block completely, or allow the user to submit anyways? `please consider another one` implies the user has a choice.

If we allow the user to submit anyways, do we make the user close the warning, check some extra box, etc, to make sure they understand the ramifications?

**Comment by shane-tomlinson** [comment]
I'm very excited for this feature @ryanfeeley!

**Comment by rfk** [comment]
@ryanfeeley IIUC you have an updated spec for this that you were working on in a Google doc - if it's ready, please update the PR to match it or link the final gdoc version here for our reference.

**Comment by rfk** [comment]
bump for @ryanfeeley to provide a link to the latest feature definition in google doc

**Comment by vladikoff** [comment]
cc @ryanfeeley @TDA , needs a feature doc

**Comment by davismtl** [comment]
@ryanfeeley we can collaborate on the feature doc if you'd like.

**Comment by rfk** [comment]
I'm going to merge the current version of this, we can iterate on it in follow-up work.
